subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsAccessFromGroovyDslIntegrationTest.groovy

```diff
@@ -258,6 +258,37 @@ class IsolatedProjectsAccessFromGroovyDslIntegrationTest extends AbstractIsolate
         "project(':').allprojects.each" | _
     }
 
+    def "reports cross-project model access on #kind lookup in the parent project using `#expr`"() {
```

Comment:
💡 It would be nice to have the same coverage for a project accessing a dynamic property two levels away. What would happen then?

Reply:
Good point! The current implementation would report two problems if the intermediate project doesn't have the property, like this:
```
2 problems were found storing the configuration cache.
- Build file 'sub/sub-sub/build.gradle': Project ':sub' cannot dynamically look up a property in the parent project ':'
- Build file 'sub/sub-sub/build.gradle': Project ':sub:sub-sub' cannot dynamically look up a property in the parent project ':sub'
```

I realise that having just one problem would be more preferable here. 
From the top of my head, there is a simple solution: introduce a build-scoped service that keeps a thread-local stacks of recursive calls to `CrossProjectModelAccessTrackingParentDynamicObject`. Each call to the functions of the wrapper class should check if it's the outermost call, and report a problem only if it is. WDYT?

Reply:
> I realise that having just one problem would be more preferable here.

👍 

> ...build-scoped service that keeps a thread-local

Or perhaps a static field in `CrossProjectModelATPDO`?

> Each call to the functions of the wrapper class should check if it's the outermost call, and report a problem only if it is. WDYT?

I fear that approach could end up hiding legitimate dynamic lookups that happen as a side-effect of the outermost call. Consider the following (unusual for sure) setup involving 3 project levels `foo` / `bar` / `baz`:

```groovy

// foo/bar/baz/build.gradle
println prop // <-- outermost lookup


// foo/bar/build.gradle
abstract class Unusual {
   Project p
   @Inject Unusual(Project p) { this.p = p }
   Object getProp() {
     println 'getProp';
     p.parent.prop // <-- this should still be reported but would happen in the context of the outermost call
   }
}

// Convention plugin members are exposed as members of the project
convention.plugins['unusual'] = objects.newInstance(Unusual)


// foo/build.gradle
ext.prop = 42 // <-- value printed
```

Reply:
I think the call to `p.parent.prop` will still be properly reported with the change I described. My idea was to only omit the reports that happen in the inner calls to the parent project dynamic objects, while `p.parent.prop` violates project isolation in a different way (accessing a property from a `ProblemReportingProject`) and thus should still trigger a problem.

However, it would still be problematic if it's not `p.parent.prop` but `p.prop` (then looked up dynamically in the parent project). 

To fix this, we could also count the calls to the "entry point" APIs (like `project.findProperty` or `project.getProperty`) in the call stack (by registering those calls in the service), and report a new problem on inherited property access whenever that counter increases. It won't increase if the parent project in turn accesses its own parent, but it will increase if some user code tries to access a property via the `Project`'s public API.

Reply:
I've added dynamic calls tracking that eliminates duplicates.